openssl/src/stack.rs

```diff
@@ -0,0 +1,312 @@
+use std::ops::{Deref, DerefMut, Index, IndexMut};
+use std::mem;
+use std::iter;
+use std::clone::Clone;
+use std::borrow::Borrow;
+use std::convert::AsRef;
+
+use libc::c_void;
+#[cfg(ossl110)]
+use libc::c_int;
+
+use ffi;
+use types::{OpenSslType, Ref};
+
+/// Trait implemented by stackable types. This must *only* be
+/// implemented on opaque types that can be directly casted into their
+/// `CType`.
+pub trait Stackable: OpenSslType {
+    /// C stack type for this element. Generally called
+    /// `stack_st_{ELEMENT_TYPE}`, normally hidden by the
+    /// `STACK_OF(ELEMENT_TYPE)` macro in the OpenSSL API.
+    type StackType;
+
+    /// When the stack is owned this destructor is used to free each
+    /// element.
+    fn destructor() -> unsafe extern fn(*mut Self::CType);
+}
+
+/// An owned stack of `T`.
+pub struct Stack<T: Stackable>(*mut T::StackType);
+
+impl<T: Stackable> Stack<T> {
+    /// Return a new Stack<T>, taking ownership of the handle
+    pub unsafe fn from_ptr(stack: *mut T::StackType) -> Stack<T> {
+        Stack(stack)
+    }
+}
+
+impl<T: Stackable> Drop for Stack<T> {
+    #[cfg(ossl10x)]
+    fn drop(&mut self) {
+        unsafe {
```

Comment:
Let's do the popping manually to avoid requiring the destructor method in Stackable:

```rust
fn drop(&mut self) {
    unsafe {
        loop {
            let ptr = OPENSSL_sk_pop(self.0 as *mut _);
            if ptr.is_null() {
                break;
            }
            T::from_ptr(ptr);
        }
        OPENSSL_sk_free(self.0 as *mut _);
    }
}
```